### PR TITLE
adding compress and take  operations

### PR DESCRIPTION
**Pull request by ipdemes**

### CODE REVIEW

File: cunumeric/module.py [hunk 1]
```diff
@@ -2323,6 +2367,60 @@ def choose(a, choices, out=None, mode="raise"):
...
+    Returns
+    -------
+    compressed_array : ndarray
+        A copy of `a` without the slices along `axis` for which condition
+        is false.
+
+    Raises
+    ------
+    ValueError : dimension mismatch
+        If condition is is not 1D array
```
Comment (manopapad) [inline cunumeric/module.py:2406]: Typo: duplicate "is"

File: cunumeric/module.py [hunk 2]
```diff
@@ -2323,6 +2367,60 @@ def choose(a, choices, out=None, mode="raise"):
...
 
 
+@add_boilerplate("c", "a")
+def compress(c, a, axis=None, out=None):
+    """
+    Return selected slices of an array along given axis.
+
+    When working along a given axis, a slice along that axis is returned
+    in output for each index where condition evaluates to True.
+    When working on a 1-D array, compress is equivalent to extract.
```
Comment (manopapad) [inline cunumeric/module.py:2377]: Should "extract" be pointing to another function?

File: cunumeric/module.py [hunk 3]
```diff
@@ -2250,6 +2250,50 @@ def where(a, x=None, y=None):
 
 
 # Indexing-like operations
+@add_boilerplate("a")
+def take(a, indices, axis=None, out=None, mode="raise"):
+    """
```
Comment (manopapad) [inline cunumeric/module.py:2255]: Can you leave empty lines between sections here?

File: cunumeric/module.py [hunk 4]
```diff
@@ -2250,6 +2250,50 @@ def where(a, x=None, y=None):
...
+    out : ndarray `(Ni…, Nj…, Nk…)`
+        The returned array has the same type as a.
+    Raises
+    ------
+    See Also
+    --------
+    numpy.take
+    Availability
+    --------
+    GPU, CPU
```
Comment (manopapad) [inline cunumeric/module.py:2294]: "Multiple GPUs, Multiple CPUs"

File: cunumeric/module.py [hunk 5]
```diff
@@ -2250,6 +2250,50 @@ def where(a, x=None, y=None):
...
+    indices : array_like `(Nj…)`
+        The indices of the values to extract.
+        Also allow scalars for indices.
+    axis : int, optional
+        The axis over which to select values. By default, the flattened input
+        array is used.
+    out : ndarray, optional `(Ni…, Nj…, Nk…)`
+        If provided, the result will be placed in this array. It should be of
+        the appropriate shape and dtype. Note that out is always buffered if
+        mode=’raise’; use other modes for better performance.
```
Comment (manopapad) [inline cunumeric/module.py:2274]: Is this relevant to our implementation?

File: cunumeric/array.py [hunk 6]
```diff
@@ -1741,6 +1741,63 @@ def astype(
...
+            # if indices is a tuple or list, bring sub-tuples to the same shape
+            # and concatenate them
+            indices = convert_to_cunumeric_ndarray(indices)
+
+        if axis is None:
+            self = self.ravel()
+            axis = 0
+        elif axis < 0:
+            axis = self.ndim + axis
+        elif axis >= self.ndim:
```
Comment (manopapad) [inline cunumeric/array.py:1755]: Move this check to a separate conditional, and check for `axis < 0` as well (even if you add `self.ndim` the value might still be negative.

File: cunumeric/array.py [hunk 7]
```diff
@@ -1741,6 +1741,63 @@ def astype(
...
+        if self.shape[axis] == 0:
+            if indices.size != 0:
+                raise IndexError(
+                    "Cannot do a non-empty take() from an empty axis."
+                )
+            return self
+
+        point_indices = tuple(slice(None) for i in range(0, axis))
+        point_indices += (indices,)
+        res = self[point_indices]
```
Comment (manopapad) [inline cunumeric/array.py:1788]: If `indices` is a scalar this will return a view, but `take` should always return a copy.

File: cunumeric/array.py [hunk 8]
```diff
@@ -1741,6 +1741,63 @@ def astype(
...
+        else:
+            raise ValueError(
+                "Invalid mode '{}' for take operation".format(mode)
+            )
+        if self.shape[axis] == 0:
+            if indices.size != 0:
+                raise IndexError(
+                    "Cannot do a non-empty take() from an empty axis."
+                )
+            return self
```
Comment (manopapad) [inline cunumeric/array.py:1784]: I would return a copy of `self` here (although it probably doesn't matter in this case).

File: cunumeric/array.py [hunk 9]
```diff
@@ -1741,6 +1741,63 @@ def astype(
...
+
+        if axis is None:
+            self = self.ravel()
+            axis = 0
+        elif axis < 0:
+            axis = self.ndim + axis
+        elif axis >= self.ndim:
+            raise ValueError("axis is out of bounds for array of dimension")
+
+        if mode == "raise":
```
Comment (manopapad) [inline cunumeric/array.py:1758]: NumPy allows negative indices in this case, e.g. for `self.shape[axis] == 2` it accepts `-1` and `-2`.

Comment (manopapad) [inline cunumeric/array.py:1758]: Since this function falls back to advanced indexing, can't we just piggyback off of the bounds checks there?

File: cunumeric/array.py [hunk 10]
```diff
@@ -1741,6 +1741,63 @@ def astype(
         result._thunk.convert(self._thunk, warn=False)
         return result
 
+    def take(self, indices, axis=None, out=None, mode="raise"):
```
Comment (manopapad) [inline cunumeric/array.py:1744]: Add a docstring for this function also.

File: cunumeric/array.py [hunk 11]
```diff
@@ -1741,6 +1741,63 @@ def astype(
...
+        point_indices += (indices,)
+        res = self[point_indices]
+        if out is not None:
+            out = convert_to_cunumeric_ndarray(out)
+            if out.shape != res.shape:
+                raise ValueError("Shape mismatch: out array has wrong shape")
+            if out.dtype != res.dtype:
+                raise ValueError("Type mismatch: out array has wrong type")
+
+            out._thunk = res._thunk
```
Comment (manopapad) [inline cunumeric/array.py:1796]: In most cases we cannot set the `._thunk` field to something else after it's been initialized, otherwise views won't work correctly (we instead update the contents of the thunk, by going through the methods in `eager.py` and `deferred.py`). In this case, if the `out` array had a view pointing to it before entry to this function, then that view would not be updated, because you are essentially creating a new region and setting `out` to it, instead of updating the contents of the region that was backing the original `out`:

```
a = cn.zeros((3,))
b = a[:]
c = cn.ones((2,3))
c.take(0, axis=0, out=a)
assert np.array_equal(a, b)
```

File: cunumeric/array.py (hunk 10 above)
Comment (manopapad) [inline cunumeric/array.py:1744]: Use `@add_boilerplate`; then you don't need to do `out = convert_to_cunumeric_ndarray(out)`.

File: cunumeric/array.py [hunk 12]
```diff
@@ -1836,6 +1893,64 @@ def choose(self, choices, out=None, mode="raise"):
         else:
             return out_arr
 
+    def compress(self, condition, axis=None, out=None):
```
Comment (manopapad) [inline cunumeric/array.py:1896]: Use `@add_boilerplate("condition")`. Then you don't need to do `out = convert_to_cunumeric_ndarray(out)`.

File: cunumeric/array.py [hunk 13]
```diff
@@ -1836,6 +1893,64 @@ def choose(self, choices, out=None, mode="raise"):
...
+        res = a[index_tuple]
+
+        if out is not None:
+            out = convert_to_cunumeric_ndarray(out)
+            if out.shape != res.shape:
+                raise ValueError("Shape mismatch: out array has wrong shape")
+            if out.dtype != res.dtype:
+                out._thunk.convert(res._thunk)
+            else:
+                out._thunk = res._thunk
```
Comment (manopapad) [inline cunumeric/array.py:1949]: As above, you cannot reset the `._thunk` field after initialization.

File: cunumeric/array.py (hunk 9 above)
Comment (ipdemes) [reply cunumeric/array.py:1758]: Advanced indexing bounds check is not implemented yet. I will add "TODO" comment before this line

File: cunumeric/array.py (hunk 11 above)
Comment (ipdemes) [reply cunumeric/array.py:1796]: should I replace it with `out._thunk.copy(res._thunk)` ?

File: cunumeric/module.py (hunk 5 above)
Comment (ipdemes) [reply cunumeric/module.py:2274]: No, it isn't

File: cunumeric/module.py (hunk 2 above)
Comment (ipdemes) [reply cunumeric/module.py:2377]: what do you mean?

Comment (manopapad) [reply cunumeric/module.py:2381]: "Extract" is a reference to another cunumeric function. Should it be pointing to the documentation for that function?

File: cunumeric/array.py (hunk 11 above)
Comment (manopapad) [reply cunumeric/array.py]: That creates an unnecessary copy, I think you could just do `out[:] = self[point_indices]`.

Comment (ipdemes) [reply cunumeric/array.py]: In this case I will perform advanced indexing operation again (which includes copy). Wouldn't it be better to call `copy`? I need to keep `res = self[point_indices]` above to check `out.shape == res.shape`.

Comment (manopapad) [reply cunumeric/array.py]: I don't think you need to create `res` if `out is not None`. `out[:] = self[point_indices]` will fail if the shape is incompatible, and you can use `self.dtype` in place of `res.dtype`.

File: cunumeric/module.py [hunk 14]
```diff
@@ -2250,6 +2250,54 @@ def where(a, x=None, y=None):
 
 
 # Indexing-like operations
+@add_boilerplate("a", "out")
```
Comment (manopapad) [inline cunumeric/module.py:2253]: "out" is handled automatically

File: cunumeric/module.py [hunk 15]
```diff
@@ -2323,6 +2371,60 @@ def choose(a, choices, out=None, mode="raise"):
     return a.choose(choices=choices, out=out, mode=mode)
 
 
+@add_boilerplate("c", "a", "out")
```
Comment (manopapad) [inline cunumeric/module.py:2374]: "out" is handled automatically.

File: cunumeric/array.py [hunk 16]
```diff
@@ -1741,6 +1741,82 @@ def astype(
...
+            # if indices is a tuple or list, bring sub-tuples to the same shape
+            # and concatenate them
+            indices = convert_to_cunumeric_ndarray(indices)
+
+        if axis is None:
+            self = self.ravel()
+            axis = 0
+        elif axis < 0:
+            axis = self.ndim + axis
+        if axis is not None and axis >= self.ndim:
```
Comment (manopapad) [inline cunumeric/array.py:1770]: `if axis < 0 or axis >= self.ndim`

Comment (ipdemes) [reply cunumeric/array.py:1770]: right, I didn't think about the case when `axis<-self.ndim`

**Comment by manopapad** [comment]
This example is failing for me:

```
import cunumeric as cn
a = cn.arange(6)
print(a.take([3], axis=0))
```

with:

```
Traceback (most recent call last):
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legion_top.py", line 410, in legion_python_main
    run_path(args[start], run_name='__main__')
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legion_top.py", line 234, in run_path
    exec(code, module.__dict__, module.__dict__)
  File "a.py", line 20, in <module>
    print(a.take([3], axis=0))
  File "/Users/mpapadakis/cunumeric/cunumeric/coverage.py", line 94, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 1815, in take
    res = self[point_indices]
  File "/Users/mpapadakis/cunumeric/cunumeric/coverage.py", line 94, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 139, in wrapper
    return func(*args, **kwargs)
  File "/Users/mpapadakis/cunumeric/cunumeric/array.py", line 794, in __getitem__
    return ndarray(shape=None, thunk=self._thunk.get_item(key))
  File "/Users/mpapadakis/cunumeric/cunumeric/deferred.py", line 639, in get_item
    copy.execute()
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legate/core/operation.py", line 183, in execute
    self._context.runtime.submit(self)
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legate/core/runtime.py", line 952, in submit
    self.flush_scheduling_window()
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legate/core/runtime.py", line 947, in flush_scheduling_window
    self._schedule(ops)
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legate/core/runtime.py", line 940, in _schedule
    op.launch(strategy)
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legate/core/operation.py", line 699, in launch
    launcher.add_output(store, req, tag=tag)
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legate/core/launcher.py", line 981, in add_output
    self.add_store(store, proj, Permission.WRITE, tag, flags)
  File "/Users/mpapadakis/legate.core/install/lib/python3.7/site-packages/legate/core/launcher.py", line 963, in add_store
    assert store.kind is not Future
AssertionError
```

**Comment by manopapad** [comment]
Negative indices don't seem to work correctly:

```
a = cn.arange(6)
print(a.take([-6,-5,-4,-3,-2,-1,0,1,2,3,4,5], axis=0))
```

prints:

```
[0 0 0 0 0 0 0 1 2 3 4 5]
```

File: cunumeric/array.py [hunk 17]
```diff
@@ -1836,6 +1912,61 @@ def choose(self, choices, out=None, mode="raise"):
...
+                category=RuntimeWarning,
+            )
+            condition = condition.astype(bool)
+        if axis is None:
+            axis = 0
+            a = self.ravel()
+
+        if a.shape[axis] < condition.shape[0]:
+            raise ValueError(
+                "Shape mismatch"
```
Comment (manopapad) [inline cunumeric/array.py:1948]: Leave a space: "Shape mismatch: "

File: cunumeric/array.py [hunk 18]
```diff
@@ -1836,6 +1912,61 @@ def choose(self, choices, out=None, mode="raise"):
...
+                slice(0, condition.shape[0]),
+            )
+            a = a[slice_tuple]
+
+        index_tuple = tuple(slice(None) for ax in range(axis))
+        index_tuple += (condition,)
+
+        if out is not None:
+            if out.dtype != self.dtype:
+                a = a._maybe_convert(out.dtype, (a, out))
```
Comment (manopapad) [inline cunumeric/array.py:1962]: I think type conversion will be handled automatically by the set_item code?

**Comment by ipdemes** [comment]
> This example is failing for me:
> 
> ```
> import cunumeric as cn
> a = cn.arange(6)
> print(a.take([3], axis=0))
> ```
> 
> with:
> 
> ```

It seems like array with the shape (1,) is considered to be a Future:
DeferredArray(base: Store(shape: Shape((1,)), ndim: None, type: int64, storage: Future(uninitialized)), transform: id)

So we can't use it in the copy operation 

I have tried t convert it to scalar by doing something like `k=key[0]`, but k is still `ndarray` type.

Any ideas?

**Comment by manopapad** [comment]
> It seems like array with the shape (1,) is considered to be a Future:
DeferredArray(base: Store(shape: Shape((1,)), ndim: None, type: int64, storage: Future(uninitialized)), transform: id)
> So we can't use it in the copy operation

^ @magnatelee ?

**Comment by magnatelee** [comment]
Stores of volume 1 in cuNumeric are all backed by futures. So @ipdemes is right that they cannot be used with copy operations currently. I can teach the core's copy how to handle futures, but in the meantime, can't you just use the copy task in cuNumeric? You can check the `kind` of the store to decide whether you can pass it to a copy operation.

**Comment by ipdemes** [comment]
> Stores of volume 1 in cuNumeric are all backed by futures. So @ipdemes is right that they cannot be used with copy operations currently. I can teach the core's copy how to handle futures, but in the meantime, can't you just use the copy task in cuNumeric? You can check the `kind` of the store to decide whether you can pass it to a copy operation.

OK, I will use copy task for the case of Future store . Thank you.

File: tests/integration/test_compress.py [hunk 19]
```diff
@@ -0,0 +1,80 @@
...
+import cunumeric as num
+from legate.core import LEGATE_MAX_DIM
+
+
+def test():
+    # test on the 1D array:
+    a = mk_seq_array(np, (10,))
+    a_num = num.array(a)
+    res = np.compress([2, 3, 6], a, axis=0)
+    res_num = num.compress([2, 3, 6], a_num, axis=0)
```
Comment (manopapad) [inline tests/integration/test_compress.py:28]: Isn't the first argument supposed to be an array of bools?

Comment (manopapad) [reply tests/integration/test_compress.py:28]: I like that this example tests the case where `len(condition) < len(array)`, I'm just wondering if it should be bools (or 0/1s instead).

**Comment by manopapad** [comment]
One last minor comment, otherwise looks good.

File: tests/integration/test_compress.py (hunk 19 above)
Comment (ipdemes) [reply tests/integration/test_compress.py]: Yes, you are right, it should be bool. The test worked because cunumeric converted it to bool (all True), there was also a warning produced

**Review by manopapad — APPROVED** [review]
